vm/actor/tests/common/mod.rs

```diff
@@ -0,0 +1,505 @@
+            .unwrap())
+    }
+
+    fn transaction<C: Cbor, R, F>(&mut self, f: F) -> Result<R, ActorError>
+    where
+        F: FnOnce(&mut C, &Self) -> R,
+    {
+        if self.in_transaction {
+            return Err(self.abort(ExitCode::SysErrorIllegalActor, "nested transaction"));
+        }
+        let mut read_only = self.state()?;
+        self.in_transaction = true;
+        let ret = f(&mut read_only, &self);
+        self.state = Some(self.put(&read_only).unwrap());
+        Ok(ret)
+    }
+
+    fn store(&self) -> &BS {
+        self.store
+    }
+
+    fn send(
+        &mut self,
+        to: &Address,
+        method: MethodNum,
+        params: &Serialized,
+        value: &TokenAmount,
+    ) -> Result<Serialized, ActorError> {
+        self.require_in_call();
+        if self.in_transaction {
+            return Err(self.abort(
+                ExitCode::SysErrorIllegalActor,
+                "side-effect within transaction",
+            ));
+        }
+
+        assert!(
+            !self.expect_sends.is_empty(),
+            "unexpected expectedMessage to: {:?} method: {:?}, value: {:?}, params: {:?}",
+            to,
+            method,
+            value,
+            params
+        );
+
+        let expected_msg = self.expect_sends[0].clone();
+
+        assert!(&expected_msg.to == to && expected_msg.method == method && &expected_msg.params == params && &expected_msg.value == value, "expectedMessage being sent does not match expectation.\nMessage -\t to: {:?} method: {:?} value: {:?} params: {:?}\nExpected -\t {:?}", to, method, value, params, &self.expect_sends[0]);
+
+        if value > &self.balance {
+            return Err(self.abort(
+                ExitCode::SysErrSenderStateInvalid,
+                format!(
+                    "cannot send value: {:?} exceeds balance: {:?}",
+                    value, self.balance
+                ),
+            ));
+        }
+        let _ = self.expect_sends.pop_front();
+        self.balance -= value;
+
+        return Ok(expected_msg.send_return);
+    }
+
+    fn abort<S: AsRef<str>>(&self, exit_code: ExitCode, msg: S) -> ActorError {
+        ActorError::new(exit_code, msg.as_ref().to_owned())
+    }
+
+    fn new_actor_address(&mut self) -> Result<Address, ActorError> {
+        self.require_in_call();
+        let ret = self
+            .new_actor_addr
+            .as_ref()
+            .expect("unexpected call to new actor address")
+            .clone();
+        self.new_actor_addr = None;
+        return Ok(ret);
+    }
+
+    fn create_actor(&mut self, code_id: &Cid, address: &Address) -> Result<(), ActorError> {
+        self.require_in_call();
+        if self.in_transaction {
+            return Err(self.abort(
+                ExitCode::SysErrorIllegalActor,
+                "side-effect within transaction".to_owned(),
+            ));
+        }
+        let expect_create_actor = self
+            .expect_create_actor
+            .clone()
```

Comment:
... Insert comment about take

Reply:
wow thats cool didnt know that method existed!